// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package osv;

import "google/protobuf/struct.proto";
import "google/protobuf/timestamp.proto";

// Commit reference.
//
// In some rare cases, this may refer to a small range of commits rather than an
// exact commit (to accomodate for automated systems). In such cases, the
// semantics are as follows:
//
// - If this is referring to a commit which introduces a vulnerability, then
// *any* commits in the range is assumed to cause the vulnerability.
//
// - If this is referring to a commit which fixes a vulnerability, then *all*
// commits in the range is assumed to cause the vulnerability.
message Commit {
  // The repository type.
  enum RepoType {
    UNSPECIFIED = 0;
    GIT = 1;
  }

  // Required. The type of the repo.
  RepoType repo_type = 1;
  // Required. The publicly accessible URL of the repo that can be directly
  // passed to clone commands.
  string repo_url = 2;
  // Required. The commit identifier (e.g. git SHA). In some cases where the
  // exact commit fails to be determined by automation, this may be a small
  // range in the form "A:B" which means the commit range from A (exclusive) to
  // B (inclusive).
  string commit = 3;
}

// Package information and version.
message Package {
  // Required. Name of the package. Should match the name used in the package
  // ecosystem (e.g. the npm package name). For C/C++ projects integrated in
  // OSS-Fuzz, this is the name used for the integration.
  string name = 1;
  // Required. The ecosystem for this package. For vulnerabilities from OSS-Fuzz
  // that do not have a package ecosystem (e.g. C/C++ packages), this is
  // "OSS-Fuzz".
  // Other valid values are:
  //   - "Go" for Go modules.
  //   - "PyPI" for Python packages.
  //   - "NPM" for NPM packages.
  string ecosystem = 2;
  // Optional. The package URL for this package.
  string purl = 3;
}

message AffectedRange {
  // Type of the version information.
  enum Type {
    UNSPECIFIED = 0;
    GIT = 1;
    SEMVER = 2;
    ECOSYSTEM = 3;
  }
  // Required. The type of version information.
  Type type = 1;
  // Required if type is GIT. The publicly accessible URL of the repo that can
  // be directly passed to clone commands.
  string repo = 2;
  // Strongly recommended. The earliest version/commit where this vulnerability
  // was introduced in. If not specified, *all* commits/versions prior to the
  // `fixed` commit/version are assumed to be affected.
  string introduced = 3;
  // Optional only if `introduced` is specified. The version/commit that this
  // vulnerability was fixed in. This must be reachable from the "introduced"
  // version/commit.
  //
  // If the vulnerability is not fixed, this will be unset.
  string fixed = 4;
}

// Affected versions and commits.
message Affects {
  // Required (at least one entry). The commit/version ranges that contain this
  // vulnerability.
  //
  // When provided, OSV will attempt to detect and append additional ranges that
  // may be affected as well (e.g. cherry-picks to other branches).
  repeated AffectedRange ranges = 1;
  // Optional. List of affected versions. This should match tag names in the
  // upstream repository. OSV will populate or add to this automatically
  // based on the provided commit ranges.
  repeated string versions = 2;
}

// Vulnerability severity information. The exact meaning of these levels are
// specific to OSS-Fuzz.
enum Severity {
  NONE = 0;
  LOW = 1;
  MEDIUM = 2;
  HIGH = 3;
  CRITICAL = 4;
}

// Reference URL.
message Reference {
  enum Type {
    NONE = 0;
    WEB = 1;
    ADVISORY = 2;
    REPORT = 3;
    FIX = 4;
    PACKAGE = 5;
  }

  // Required. The type of the reference.
  Type type = 1;
  // Required. The URL.
  string url = 2;
}

// A vulnerability entry.
// The protobuf representation is *NOT* stable and only used for implementing
// the JSON based API.
message Vulnerability {
  // The `id` field is a unique identifier for the vulnerability entry. It is a
  // string of the format `<DB>-<ENTRYID>`, where `DB` names the database and
  // `ENTRYID` is in the format used by the database. For example:
  // “OSV-2020-111”, “CVE-2021-3114”, or “GHSA-vp9c-fpxx-744v”.
  string id = 1;
  // The RFC3339 timestamp indicating when this entry was published.
  google.protobuf.Timestamp published = 11;
  // The RFC3339 timestamp indicating when this entry was last modified.
  google.protobuf.Timestamp modified = 10;
  // Optional. The RFC3339 timestamp indicating when this entry is considered to
  // be withdrawn.
  google.protobuf.Timestamp withdrawn = 12;
  // Optional. IDs for the same vulnerability in other databases.
  repeated string aliases = 8;
  // Optional. List of IDs of closely related vulnerabilities, such as the same
  // problem in alternate ecosystems.
  repeated string related = 13;
  // Required. Package information.
  Package package = 2;
  // Required. One line human readable summary for the vulnerability. It is
  // recommended to keep this under 120 characters.
  string summary = 3;
  // Required. Any additional human readable details for the vulnerability.
  string details = 4;
  // Required. Affected commit ranges and versions.
  Affects affects = 6;
  // Optional. URLs to more information/advisories (including the
  // scheme e.g "https://").
  repeated Reference references = 16;
  // Optional. JSON object holding additional information about the
  // vulnerability as defined by the ecosystem for which the record applies.
  google.protobuf.Struct ecosystem_specific = 14;
  // Optional. JSON object holding additional information about the
  // vulnerability as defined by the database for which the record applies.
  google.protobuf.Struct database_specific = 15;
  // Deprecated and will be removed in the future.
  Severity severity = 5;
}
